[Docs] dpl documentation (#3700)
The-OpenROAD-Project/OpenROAD: src/dpl/README.md

## Commands

### Detailed Placement

The `detailed_placement` command performs detailed placement of instances
to legal locations after global placement.

```tcl
detailed_placement
    [-max_displacement disp|{disp_x disp_y}]

>     [-disallow_one_site_gaps]

Reviewer: Missing `-report_file_name`

```

#### Options

| Switch Name | Description | 
| ----- | ----- |

> | `-max_displacement` | How far an instance (in microns) can be moved when finding a site where it can be placed. Either set one value for both directions or set {disp_x disp_y} for individual directions. (default 0, 0, integer). |

Reviewer: ```suggestion
| `-max_displacement` | Max distance that an instance can be moved (in microns) when finding a site where it can be placed. Either set one value for both directions or set `{disp_x disp_y}` for individual directions. (default `{0, 0}`, integer). |
```

specify padding for all instances of a common master, use the `-filter`
"ref_name == <name>" option to `get_cells`.

```tcl
set_placement_padding   
    -global|-masters masters|-instances insts
    [-right site_count]
    [-left site_count]

>     [instances]

Reviewer: Please add a description for the last `[instances]` and clearly differentiate `-instances vs [instances]`.

Author: Should we change the name of the `-instances` switch then?

Reviewer: @mina1460
Would it be correct to remove `[instances]`? Looking at the code the below looks correct to me.

```suggestion
    -global|-masters masters|-instances insts
    [-right site_count]
    [-left site_count]
```

```

#### Options

| Switch Name | Description | 
| ----- | ----- |

> | `-global` or `-masters` or `-instances` | Either flag must be set. For `-global`, you will set padding globally using `left` and `right` values. For `-masters`, you will set padding only for these masters using `left` and `right` values. For `-instances`, you will set padding only for these insts using `left` and `right` values. |

Reviewer: Explain each in their own row. Add a warning callout before the table to explain that only one of the three will be honored, and mention that the order of preference is `global > masters > instances`. If you need clarification, please ask mina1460.

> | `left` | Left padding (in site count). |
> | `right` | Right padding (in site count). |

Reviewer: ```suggestion
| `-left` | Left padding (in site count). |
| `-right` | Right padding (in site count). |
```

FILLCELL_X32 FILLCELL_X4 FILLCELL_X8`.  To specify a different naming prefix
from `FILLER_` use `-prefix <new prefix>`.

```tcl 
filler_placement
    [-prefix prefix]
    filler_masters
```

#### Options

| Switch Name | Description |
| ----- | ----- |
| `-prefix` | prefix to name the filler cells. (default "FILLER_") |
| `filler_masters` | filler master cells | 

### Remove Fillers

```tcl
remove_fillers 
```

No arguments are needed for this function. 

### Check Placement

The `check_placement` command checks the placement legality. It returns
`0` if the placement is legal.

```tcl
check_placement
    [-verbose]
    [-disallow_one_site_gaps]
```

#### Options

| Switch Name | Description |
| ----- | ----- |
| `-verbose` | enable verbose logging |
| `-disallow_one_site_gaps` | disable one site gaps during placement |

### Optimize Mirroring

The `optimize_mirroring` command mirrors instances about the Y axis in
a weak attempt to reduce total wirelength (HPWL).

```tcl
optimize_mirroring
```

No arguments are needed for this function. 

## Example scripts

> Examples scripts demonstrating how to run OpenDP on a sample design of `aes` as follows:

Reviewer: Re-opening as it is still present.

FILLCELL_X32 FILLCELL_X4 FILLCELL_X8`.  To specify a different naming prefix
from `FILLER_` use `-prefix <new prefix>`.

```tcl 
filler_placement
    [-prefix prefix]
    filler_masters
```

#### Options

| Switch Name | Description |
| ----- | ----- |
| `-prefix` | Prefix to name the filler cells (default "FILLER_"). |
| `filler_masters` | Filler master cells. | 

### Remove Fillers

> 

Reviewer: This command removes all filler cells.

```tcl
remove_fillers 
```

No arguments are needed for this function. 

### Check Placement

The `check_placement` command checks the placement legality. It returns
`0` if the placement is legal.

```tcl
check_placement
    [-verbose]

>     [-disallow_one_site_gaps]

Reviewer: Missing `-report_filename`

```

#### Options

| Switch Name | Description |
| ----- | ----- |
| `-verbose` | Enable verbose logging. |

> | `-disallow_one_site_gaps` | Disable one site gaps during placement. |

Reviewer: ```suggestion
| `-disallow_one_site_gaps` | Disable one site gap during placement check. |
```

Reviewer: A note about using the same as used in `detailed_placement` to guarantee consistency would be nice.

### Optimize Mirroring

> The `optimize_mirroring` command mirrors instances about the Y axis in
> a weak attempt to reduce total half-perimeter wirelength (HPWL).

Reviewer: ```suggestion
The `optimize_mirroring` command mirrors instances about the Y axis in
a weak attempt to reduce the total half-perimeter wirelength (HPWL).
```

```tcl
optimize_mirroring
```

No arguments are needed for this function. 

### Useful developer functions

If you are a developer, you might find these useful. More details can be found in the [source file](./src/Opendp.cpp) or the [swig file](./src/Opendp.i).

```tcl
# debug detailed placement
detailed_placement_debug 
    -min_displacement disp
    -instance inst

# get masters from a design
get_masters_arg

# get bounding box of an instance
get_inst_bbox inst_name

# get grid bounding box of an instance
get_inst_grid_bbox inst_name

# format grid (takes in length x and site width w)
format_grid x w

# get row site name
get_row_site

> ```

Reviewer: Convert to table.

## Example scripts

Examples scripts demonstrating how to run OpenDP on a sample design of `aes` as follows:

```shell
./test/aes.tcl
```

## Regression tests

> There are a set of regression tests in `/test`. For more information, refer to this [section](../../README.md#regression-tests). 

Reviewer: ```suggestion
There are a set of regression tests in `./test/`. Refer to this [section](../../README.md#regression-tests) for more information. 
```
